test/python/test_jacobian.py

```diff
@@ -432,683 +496,1022 @@ def calc(mode):
+        gas.TPX = 300, 2 * ct.one_atm, "H2:1, O2:3, AR:0.4"
+        return gas
 
 
-class FullEdgeCases(FullTests, utilities.CanteraTest):
+class SurfaceRateExpressionTests:
+    """
+    Generic test class to check derivatives evaluated for a single reaction within
+    a reaction mechanism for surfaces.
+    """
+    rtol = 1e-5  # Relative tolerance for approximate comparisons
+
+    @pytest.fixture(scope='class')
+    def all_species(self, gas, surf):
+        """Combine species from both gas and surface phases."""
+        return surf.species() + gas.species()
+
+    @pytest.fixture(scope='class')
+    def sidxs(self, all_species):
+        """Create a dictionary mapping species names to indices."""
+        return {spec.name: i for i, spec in enumerate(all_species)}
+
+    @pytest.fixture(scope='class')
+    def rxn_data(self, surf, sidxs, rxn_idx):
+        """
+        Retrieve reaction-specific data based on the reaction index.
+        Returns a dictionary containing reaction, reactant indices, and product indices.
+        """
+        rxn = surf.reactions()[rxn_idx]
+        r_stoich = surf.reactant_stoich_coeffs
+        p_stoich = surf.product_stoich_coeffs
+        rix = [sidxs[k] for k in rxn.reactants.keys()]
+        pix = [sidxs[k] for k in rxn.products.keys()]
+        return {
+            "rxn": rxn,
+            "r_stoich": r_stoich,
+            "p_stoich": p_stoich,
+            "rix": rix,
+            "pix": pix
+        }
+
+    @pytest.fixture(scope='class')
+    def initial_conditions(self, gas, surf):
+        """Store initial TPX conditions for gas and surface."""
+        return {
+            "gas_tpx": gas.TPX,
+            "surf_tpx": surf.TPX
+        }
+
+    @pytest.fixture(scope='function', autouse=True)
+    def setup_surface_rate_expression_data(self, gas, surf, initial_conditions, rxn_data, rxn_idx, sidxs):
+        """
+        Reset TPX and derivative settings before each test and configure reaction multipliers.
+        Also verify stoichiometric coefficients.
+        """
+        # Reset gas phase
+        gas.TPX = initial_conditions["gas_tpx"]
+        gas.set_multiplier(0)
+        gas.derivative_settings = {}  # Reset to defaults
+
+        # Reset surface phase
+        surf.TPX = initial_conditions["surf_tpx"]
+        surf.set_multiplier(0.0)
+        surf.set_multiplier(1.0, rxn_idx)  # Enable multiplier for specific reaction
+        surf.derivative_settings = {
+            "skip-coverage-dependence": True,
+            "skip-electrochemistry": True
+        }
+
+        # Retrieve reaction data
+        rxn = rxn_data["rxn"]
+        r_stoich = rxn_data["r_stoich"]
+        p_stoich = rxn_data["p_stoich"]
+
+        # Check stoichiometric coefficients for reactants
+        for k, v in rxn.reactants.items():
+            ix = sidxs[k]
+            actual = r_stoich[ix, rxn_idx]
+            assert actual == v, f"Reactant stoich mismatch for species '{k}': expected {v}, got {actual}"
```

Comment:
Pytest already does a good job of showing you the actual / expected values, so extra message here (and in several other tests) can be simplified:
```suggestion
        for k, v in rxn.reactants.items():
            ix = sidxs[k]
            actual = r_stoich[ix, rxn_idx]
            assert actual == v, f"Reactant stoich mismatch for species '{k}'"
```